flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java

```diff
@@ -261,7 +298,9 @@ public void invoke(IN next) throws Exception {
 		} else {
 			record = new ProducerRecord<>(targetTopic, partitioner.partition(next, serializedKey, serializedValue, partitions.length), serializedKey, serializedValue);
 		}
-
+		if(flushOnCheckpoint) {
+			pendingRecords++;
```

Comment:
The increment and decrement operations are not atomic. Thus, we have to guard them from concurrent access. Concurrent modifications can happen when you're in the `invoke` method and a callback is executed at the same time.